### DISCUSSION TITLE
Prepared Transaction Mock

### DISCUSSION THREAD

**Discussion by jonmchan** [hyperledger/fabric-gateway#358] [General]
The prepared transaction is extremely complex. The fact that many of the objects are serialized instead of embedded directly makes it hard to trace down the hierarchy without having to refer to the protos constantly. I had to create my own mocked prepared transaction messages for testing, thought I'd leave this here if anyone found them useful. Unfortunately, it is in ruby, but it shouldn't be too hard to convert it to your language of choice.

```ruby
Gateway::PreparedTransaction.new(
  transaction_id: 'some_id',
  envelope: Common::Envelope.new(
    payload: Common::Payload.new(
      header: Common::Header.new(
        channel_header: Common::ChannelHeader.new(
          channel_id: 'channel_id'
        ).to_proto
      ),
      data: Protos::Transaction.new(actions:
        [
          Protos::TransactionAction.new(
            payload: Protos::ChaincodeActionPayload.new(
              action: Protos::ChaincodeEndorsedAction.new(
                proposal_response_payload: Protos::ProposalResponsePayload.new(
                  extension: Protos::ChaincodeAction.new(
                    response: Protos::Response.new(
                      status: 200,
                      message: '',
                      payload: 'first'
                    )
                  ).to_proto
                ).to_proto
              )
            ).to_proto
          )
        ]).to_proto
    ).to_proto,
    signature: 'signature'
  )
)
```

**Comment by bestbeforetoday** [comment]
Some of the protobuf structures certainly are very complex -- maybe overly complex -- particularly where there are multiple nested layers of serialized protobuf messages. Unfortunately this is now a difficult thing to change.

For reference, there are similar mock messages generated in Go, Node and Java unit tests:

- https://github.com/hyperledger/fabric-gateway/blob/1e4a926ddb98ec8ee969da3fc1500642ab389d01/pkg/client/submit_test.go#L32-L61
- https://github.com/hyperledger/fabric-gateway/blob/1e4a926ddb98ec8ee969da3fc1500642ab389d01/node/src/client.test.ts#L176-L219
- https://github.com/hyperledger/fabric-gateway/blob/1e4a926ddb98ec8ee969da3fc1500642ab389d01/java/src/test/java/org/hyperledger/fabric/client/TestUtils.java#L106-L146

**Comment by jonmchan** [comment]
Thanks for pointing that out. The 12 indents 8 spaces per tab of the go test makes me want to cry. 😅  😂

**Reply by bestbeforetoday** [reply]
Welcome to the wonderful world of [Go code formatting](https://go.dev/doc/effective_go#formatting)! In fairness, the Go indentation is all tabs so you can set your editor to render them as whatever number of spaces makes you happy.
